presto-rcfile/src/main/java/com/facebook/presto/rcfile/RcFileReader.java

```diff
@@ -0,0 +1,584 @@
+
+        // read header
+        Slice magic = input.readSlice(RCFILE_MAGIC.length());
+        boolean compressed;
+        if (RCFILE_MAGIC.equals(magic)) {
+            version = input.readByte();
+            if (version > CURRENT_VERSION) {
+                throw corrupt("RCFile version %s not supported: %s", version, dataSource);
+            }
+
+            compressed = input.readBoolean();
+        }
+        else if (SEQUENCE_FILE_MAGIC.equals(magic)) {
+            // first version of RCFile used magic SEQ with version 6
+            byte sequenceFileVersion = input.readByte();
+            if (sequenceFileVersion == SEQUENCE_FILE_VERSION) {
+                throw corrupt("File %s is a SequenceFile not an RCFile", dataSource);
+            }
+
+            // this is the first version of RCFile
+            this.version = FIRST_VERSION;
+
+            Slice keyClassName = RcFileDecoderUtils.readLengthPrefixedString(input);
+            Slice valueClassName = RcFileDecoderUtils.readLengthPrefixedString(input);
+            if (!RCFILE_KEY_BUFFER_NAME.equals(keyClassName) || !RCFILE_VALUE_BUFFER_NAME.equals(valueClassName)) {
+                throw corrupt("File %s is a SequenceFile not an RCFile", dataSource);
+            }
+
+            compressed = input.readBoolean();
+
+            // RC file is never block compressed
+            if (input.readBoolean()) {
+                throw corrupt("File %s is a SequenceFile not an RCFile", dataSource);
+            }
+        }
+        else {
+            throw corrupt("File %s is not an RCFile", dataSource);
+        }
+
+        // setup the compression codec
+        if (compressed) {
+            Slice codecClassName = RcFileDecoderUtils.readLengthPrefixedString(input);
+            this.codec = codecFactory.createCompressionCodec(codecClassName.toStringUtf8());
+        }
+        else {
+            this.codec = null;
+        }
+
+        // read metadata
+        int metadataEntries = Integer.reverseBytes(input.readInt());
+        if (metadataEntries < 0) {
+            throw corrupt("Invalid metadata entry count %s in RCFile %s", metadataEntries, dataSource);
+        }
+        ImmutableMap.Builder<String, String> metadataBuilder = ImmutableMap.builder();
+        for (int i = 0; i < metadataEntries; i++) {
+            metadataBuilder.put(RcFileDecoderUtils.readLengthPrefixedString(input).toStringUtf8(), RcFileDecoderUtils.readLengthPrefixedString(input).toStringUtf8());
+        }
+        metadata = metadataBuilder.build();
+
+        // get column count from metadata
+        String columnCountString = metadata.get(COLUMN_COUNT_METADATA_KEY);
+        try {
+            columnCount = Integer.parseInt(columnCountString);
+        }
+        catch (Exception e) {
+            throw corrupt("Invalid column count %s in RCFile %s", columnCountString, dataSource);
+        }
+
+        // initialize columns
+        columns = new Column[columnCount];
```

Comment:
In general, should we sanity check things like `columnCount` or `metadataEntries`?